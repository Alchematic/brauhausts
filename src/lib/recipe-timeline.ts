import convert from 'convert-units';
import * as _ from 'lodash';
import { Fermentable } from './fermentable';
import { GLOBALS } from './globals';
import { computeMashStepDescription, createMash, Mash, MashStep } from './mash';
import { computeRecipeGrainWeight, Recipe } from './recipe';
import { Spice } from './spice';
import { computeDisplayDuration, computeTimeToCool, computeTimeToHeat, convertKgToLbOz } from './utils';
import { Yeast } from './yeast';

export type TimelineMap = {
  fermentables: TimelineFermentables;
  times: { [key: number]: TimelineSpice[] };
  drySpice: { [key: number]: TimelineSpice[] };
  yeast: Yeast[];
};

type TimelineFermentables = {
  mash: TimelineFermentable[];
  steep: TimelineFermentable[];
  boil: TimelineFermentable[];
  boilEnd: TimelineFermentable[];
};

type TimelineFermentable = {
  fermentable: Fermentable;
  gravity: number;
};

type TimelineSpice = {
  spice: Spice;
  bitterness: number;
};

type BrewState = {
  timeline: { time: number; instructions: string }[];
  volume: number;
  temp: number;
  time: number;
  isSiUnits: boolean;
};

/**
 * Compute the number of servings a recipe can create
 */
const computeRecipeServings = (recipe: Recipe) => Math.floor(recipe.batchSize / recipe.servingSize);

/**
 * Get a list of fermentable descriptions taking siUnits into account
 */
const createFermentableIngredientList = (fermentables: TimelineFermentable[], isSiUnits = true) =>
  _.map(fermentables, ({ fermentable, gravity }) => {
    const weight = isSiUnits ? `${fermentable.weight.toFixed(2)}kg` : convertKgToLbOz(fermentable.weight);

    return `${weight} of ${fermentable.name} (${gravity.toFixed(1)} GU)`;
  });

/**
 * Get a list of spice descriptions taking siUnits into account
 */
const createSpiceIngredientList = (spices: TimelineSpice[], isSiUnits = true) =>
  _.map(spices, ({ spice, bitterness }) => {
    const weight = isSiUnits ? `${spice.weight * 1000}g` : convertKgToLbOz(spice.weight);
    const ibu = bitterness ? ` (${bitterness.toFixed(1)} IBU)` : '';

    return `${weight} of ${spice.name}${ibu}`;
  });

const generateMashStepVolumeAdd = (
  step: Readonly<MashStep>,
  recipe: Readonly<Recipe>,
  strikeVolume: number,
  currentState: BrewState,
): BrewState => {
  // We are adding hot or cold water!
  // 4.184 is the specific heat of water. Not sure what's being computed here.
  const strikeTemp =
    ((step.temp - currentState.temp) * ((GLOBALS.SPECIFIC_HEAT_OF_WATER / 10) * computeRecipeGrainWeight(recipe))) /
      strikeVolume +
    step.temp;
  const timeToHeat = computeTimeToHeat(strikeVolume, strikeTemp - currentState.temp);

  const strikeVolumeDesc = currentState.isSiUnits
    ? `${strikeVolume.toFixed(1)}l`
    : `${convert(strikeVolume)
        .from('l')
        .to('qt')}qts`;

  const strikeTempDesc = currentState.isSiUnits
    ? `${Math.round(strikeTemp)}°C`
    : `${Math.round(
        convert(strikeTemp)
          .from('C')
          .to('F'),
      )}°F`;

  return {
    ...currentState,
    timeline: _.concat(currentState.timeline, {
      time: currentState.time,
      instructions: `Heat ${strikeVolumeDesc} to ${strikeTempDesc} (about ${Math.round(timeToHeat)} minutes)`,
    }),
    time: timeToHeat + currentState.time,
    volume: strikeVolume + currentState.volume,
  };
};

const generateMashStepHeat = (step: MashStep, currentState: BrewState) => {
  let heatTemp = '';
  const timeToHeat = computeTimeToHeat(currentState.volume, step.temp - currentState.temp);

  if (currentState.isSiUnits) {
    heatTemp = `${Math.round(step.temp)}°C`;
  } else {
    heatTemp = `${Math.round(
      convert(step.temp)
        .from('C')
        .to('F'),
    )}°F`;
  }

  return {
    ...currentState,
    timeline: _.concat(currentState.timeline, {
      time: currentState.time,
      instructions: `Heat the mash to ${heatTemp} (about ${Math.round(timeToHeat)} minutes)`,
    }),
    time: currentState.time + timeToHeat,
  };
};

const computeMashPhase = (recipe: Readonly<Recipe>, currentState: BrewState) => {
  const mash = createMash(recipe, currentState.temp, recipe.mash);

  const ingredients = createFermentableIngredientList(recipe.timelineMap.fermentables.mash);
  currentState.timeline.push({
    time: currentState.time,
    instructions: `Begin ${mash.name} mash. Add ${ingredients.join(', ')}.`,
  });

  const steps = recipe.mash.steps;

  _.each(steps, step => {
    const strikeVolume = step.waterRatio * computeRecipeGrainWeight(recipe) - currentState.volume;

    if (step.temp !== currentState.temp && strikeVolume > 0) {
      currentState = generateMashStepVolumeAdd(step, recipe, strikeVolume, currentState);
    } else if (step.temp !== currentState.temp) {
      currentState = generateMashStepHeat(step, currentState);
    }

    currentState.timeline.push({
      time: currentState.time,
      instructions: `${step.name}: ${computeMashStepDescription(
        step,
        currentState.isSiUnits,
        computeRecipeGrainWeight(recipe),
      )}.`,
    });
    currentState.time += step.time;
    currentState.temp = step.temp - (step.time * GLOBALS.MASH_HEAT_LOSS) / 60.0;
  });

  currentState.timeline.push({
    time: currentState.time,
    instructions: 'Remove grains from mash. This is now your wort.',
  });
  currentState.time += 5;

  return currentState;
};

const computeSteepPhase = (recipe: Readonly<Recipe>, currentState: BrewState) => {
  let steepWeight = 0;

  _.each(recipe.timelineMap.fermentables.steep, ({ fermentable }) => {
    steepWeight += fermentable.weight;
  });

  const STEEP_LITERS_PER_KG = 2.75;
  const STEEP_TEMP = 68;
  const steepVolume = steepWeight * STEEP_LITERS_PER_KG;

  const steepHeatTime = computeTimeToHeat(steepVolume, STEEP_TEMP - currentState.temp);
  currentState.temp = STEEP_TEMP;
  currentState.volume += steepVolume;

  const steepVolumeText = currentState.isSiUnits
    ? `${steepVolume.toFixed(1)}l`
    : `${convert(steepVolume)
        .from('l')
        .to('gal')
        .toFixed(1)}gal`;
  const steepTemp = currentState.isSiUnits
    ? `${STEEP_TEMP}°C`
    : `${convert(STEEP_TEMP)
        .from('C')
        .to('F')
        .toFixed(1)}°F`;

  currentState.timeline.push({
    time: currentState.time,
    instructions: `Heat ${steepVolumeText} to ${steepTemp} (about ${Math.round(steepHeatTime)} minutes)`,
  });
  currentState.time += steepHeatTime;

  const ingredients = createFermentableIngredientList(recipe.timelineMap.fermentables.steep);
  currentState.timeline.push({
    time: currentState.time,
    instructions: `Add ${ingredients.join(', ')} and steep for ${recipe.steepTime} minutes.`,
  });
  currentState.time += recipe.steepTime;

  return currentState;
};

const computeTopUpStep = (recipe: Recipe, currentState: BrewState, boilName: string) => {
  // Adjust temperature based on added water
  const waterChangeRatio = Math.min(1, currentState.volume / recipe.boilSize);
  currentState.temp = currentState.temp * waterChangeRatio + GLOBALS.ROOM_TEMP * (1.0 - waterChangeRatio);

  const boilVolume = currentState.isSiUnits
    ? `${recipe.boilSize.toFixed(1)}l`
    : `${convert(recipe.boilSize)
        .from('l')
        .to('gal')
        .toFixed(1)}gal`;

  // Old: recipe.boilSize - currentState.volume < recipe.boilSize
  // ^ That's equivalent to currentState.volume > 0
  const action =
    currentState.volume > 0
      ? `Top up the ${boilName} to ${boilVolume} and heat to a rolling boil`
      : `Bring ${boilVolume} to a rolling boil`;

  const boilTime = computeTimeToHeat(recipe.boilSize, 100 - currentState.temp);
  currentState.timeline.push({ time: currentState.time, instructions: `${action} (about ${boilTime} minutes).` });
  currentState.time += boilTime;

  return currentState;
};

const computeBoilPhase = (recipe: Recipe, currentState: BrewState) => {
  const times = _.orderBy(Object.keys(recipe.timelineMap.times), _.parseInt, 'desc');

  // If we have late additions and no late addition time, add it
  if (!_.isEmpty(recipe.timelineMap.fermentables.boilEnd) && !_.includes(times, '5')) {
    recipe.timelineMap.times[5] = [];
    times.push('5');
  }

  // Add fermentables that need to be boiled for a certain amount of time.
  let previousSpiceTime = 0;
  _.each(times, (unparsedTime, index) => {
    const time = parseInt(unparsedTime);
    let ingredients = createSpiceIngredientList(recipe.timelineMap.times[time]);

    if (index === 0) {
      const boilIngredients = createFermentableIngredientList(recipe.timelineMap.fermentables.boil);
      ingredients = _.concat(boilIngredients, ingredients);
      previousSpiceTime = time;
    }

    currentState.time += previousSpiceTime - time;

    previousSpiceTime = time;

    if (time === 5 && recipe.timelineMap.fermentables.boilEnd.length) {
      const boilEndIngredients = createFermentableIngredientList(recipe.timelineMap.fermentables.boilEnd);
      ingredients = _.concat(boilEndIngredients, ingredients);
    }

    currentState.timeline.push({ time: currentState.time, instructions: `Add ${ingredients.join(', ')}` });
  });

  currentState.time += previousSpiceTime;

  return currentState;
};

/**
 * Compute a recipe's timeline of instructions
 */
export const computeRecipeTimeline = (recipe: Readonly<Recipe>, isSiUnits = true) => {
  let boilName = 'water';
  let currentState: BrewState = {
    timeline: [],
    time: 0,
    temp: GLOBALS.ROOM_TEMP,
    volume: 0,
    isSiUnits: isSiUnits,
  };

  if (_.size(recipe.timelineMap.fermentables.mash) > 0) {
    boilName = 'wort';
    currentState = computeMashPhase(recipe, currentState);
  }

  if (recipe.timelineMap.fermentables.steep.length) {
    boilName = 'wort';
    currentState = computeSteepPhase(recipe, currentState);
  }

  currentState = computeTopUpStep(recipe, currentState, boilName);

  // Removed mutation
  // recipe.boilStartTime = currentState.time;

  currentState = computeBoilPhase(recipe, currentState);

  // Removed mutation
  // recipe.boilEndTime = currentState.time;

  const chillTemp = isSiUnits
    ? `${recipe.primaryTemp}°C`
    : `${convert(recipe.primaryTemp)
        .from('C')
        .to('F')}°F`;

  computeTimeToCool(currentState.volume, 100, recipe.primaryTemp);
  currentState.timeline.push({
    time: currentState.time,
    instructions: `Flame out. Begin chilling to ${chillTemp} and aerate the cooled wort (about 20 minutes).`,
  });
  currentState.time += 20;

  let yeasts = _.map(recipe.yeast, 'name');

  if (_.isEmpty(yeasts) && recipe.primaryDays) {
    // No yeast given, but primary fermentation should happen...
    // Let's just use a generic "yeast" to pitch.
    yeasts = ['yeast'];
  }

  if (yeasts.length) {
    currentState.timeline.push({
      time: currentState.time,
      instructions: `Pitch ${yeasts.join(
        ', ',
      )} and seal the fermenter. You should see bubbles in the airlock within 24 hours.`,
    });
  }

  // The brew day is over! Fermenting starts now.
  // Removed mutation
  // recipe.brewDayDuration = currentState.time;

  if (!recipe.primaryDays && !recipe.secondaryDays && !recipe.tertiaryDays) {
    currentState.timeline.push({
      time: currentState.time,
      instructions: `Drink immediately (about ${computeRecipeServings(recipe)} bottles).`,
    });

    return currentState.timeline;
  }

  currentState.time += recipe.primaryDays * 1440;

  if (recipe.secondaryDays) {
    currentState.timeline.push({
      time: currentState.time,
      instructions: `Move to secondary fermenter for ${computeDisplayDuration(recipe.secondaryDays * 1440, 2)}.`,
    });
    currentState.time += recipe.secondaryDays * 1440;
  }
  if (recipe.tertiaryDays) {
    currentState.timeline.push({
      time: currentState.time,
      instructions: `Move to tertiary fermenter for ${computeDisplayDuration(recipe.tertiaryDays * 1440, 2)}.`,
    });
    currentState.time += recipe.tertiaryDays * 1440;
  }
  let primeMsg = `Prime and bottle about ${computeRecipeServings(recipe)} bottles.`;

  if (recipe.agingDays) {
    let ageTemp = '';
    if (isSiUnits) {
      ageTemp = `${recipe.agingTemp}C`;
    } else {
      ageTemp = `${convert(recipe.agingTemp)
        .from('C')
        .to('F')}F`;
    }

    primeMsg += ` Age at ${ageTemp} for ${recipe.agingDays} days.`;
  }
  currentState.timeline.push({ time: currentState.time, instructions: primeMsg });
  currentState.time += recipe.agingDays * 1440;

  currentState.timeline.push({ time: currentState.time, instructions: "Relax, don't worry and have a homebrew!" });

  return currentState.timeline;
};
