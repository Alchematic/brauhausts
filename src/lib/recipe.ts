import convert from 'convert-units';
import * as _ from 'lodash';
import { computeFermentableGU, computeFermentablePrice, computeFermentableUse, Fermentable } from './fermentable';
import { GLOBALS } from './globals';
import { Mash } from './mash';
import { TimelineFermentable, TimelineMap } from './recipe-timeline';
import { computeIsDrySpice, computeSpiceBitterness, computeSpicePrice, Spice } from './spice';
import { Style } from './style';
import { convertSpecificGravityToPlato, lowerCaseIncludes, mergeObjects } from './utils';
import { computeYeastPrice, Yeast } from './yeast';
export { importBeerXML } from './import-beerxml';

export enum RecipeType {
  EXTRACT = 'Extract',
  PARTIAL_MASH = 'Partial Mash',
  ALL_GRAIN = 'All Grain',
}

export const RecipeTypeMap = {
  Extract: RecipeType.EXTRACT,
  'Partial Mash': RecipeType.PARTIAL_MASH,
  'All Grain': RecipeType.ALL_GRAIN,
};

/**
 * A beer recipe, consisting of various ingredients and metadata which
 * provides a calculate() method to calculate OG, FG, IBU, ABV, and a
 * timeline of instructions for brewing the recipe.
 */
export type Recipe = {
  name: string;
  description: string;
  type: RecipeType;
  author: string;
  notes: string;
  boilSize: number;
  batchSize: number;
  servingSize: number;

  steepEfficiency: number;
  steepTime: number;
  mashEfficiency: number;

  style: Style;
  ibuMethod: 'tinseth' | 'rager';

  fermentables: Fermentable[];
  spices: Spice[];
  yeast: Yeast[];

  mash: Mash;

  og: number;
  est_og: number;
  fg: number;
  est_fg: number;
  color: number;
  est_color: number;
  ibu: number;
  abv: number;
  est_abv: number;
  price: number;

  buToGu: number; // Bitterness to gravity ratio
  bv: number; // Balance value (http://klugscheisserbrauerei.wordpress.com/beer-balance/)

  ogPlato: number;
  fgPlato: number;
  abw: number;
  realExtract: number;
  calories: number;

  bottlingTemp: number;
  bottlingPressure: number;

  kegTemp: number;
  kegPressure: number;

  primingCornSugar: number;
  primingSugar: number;
  primingHoney: number;
  primingDme: number;

  primaryDays: number;
  primaryTemp: number;
  secondaryDays: number;
  secondaryTemp: number;
  tertiaryDays: number;
  tertiaryTemp: number;
  agingDays: number;
  agingTemp: number;

  brewDayDuration?: any;
  boilStartTime?: any;
  boilEndTime?: any;

  timelineMap?: TimelineMap; // A mapping of values used to build a recipe timeline / instructions
};

export const createRecipe = (overrideRecipe?: Partial<Recipe>): Recipe => {
  const newRecipe: Recipe = {
    name: 'New Recipe',
    description: 'Recipe description',
    type: RecipeType.EXTRACT,
    author: 'Anonymous Brewer',
    notes: '',
    boilSize: 10.0,
    batchSize: 20.0,
    servingSize: 0.355,
    steepEfficiency: 50,
    steepTime: 20,
    mashEfficiency: 75,
    style: {
      name: '',
      category: '',
      og: [1.0, 1.15],
      fg: [1.0, 1.15],
      ibu: [0, 150],
      color: [0, 500],
      abv: [0, 14],
      carb: [1.0, 4.0],
    },
    ibuMethod: 'tinseth',
    fermentables: [],
    spices: [],
    yeast: [],
    mash: null,
    og: 0.0,
    est_og: 0.0,
    fg: 0.0,
    est_fg: 0.0,
    color: 0.0,
    est_color: 0.0,
    ibu: 0.0,
    abv: 0.0,
    est_abv: 0.0,
    price: 0.0,
    buToGu: 0.0,
    bv: 0.0,
    ogPlato: 0.0,
    fgPlato: 0.0,
    abw: 0.0,
    realExtract: 0.0,
    calories: 0.0,
    bottlingTemp: 0.0,
    bottlingPressure: 0.0,
    kegTemp: 0.0,
    kegPressure: 0.0,
    primingCornSugar: 0.0,
    primingSugar: 0.0,
    primingHoney: 0.0,
    primingDme: 0.0,
    primaryDays: 14.0,
    primaryTemp: 20.0,
    secondaryDays: 0.0,
    secondaryTemp: 0.0,
    tertiaryDays: 0.0,
    tertiaryTemp: 0.0,
    agingDays: 14,
    agingTemp: 20.0,
  };

  return _.defaultsDeep(overrideRecipe, newRecipe);
};

export const addToRecipe = (recipe: Recipe, type: 'fermentable' | 'spice' | 'hop' | 'yeast', values: any) => {
  const newRecipe = _.cloneDeep(recipe);
  switch (type) {
    case 'fermentable':
      newRecipe.fermentables.push(values);
      break;
    case 'spice':
    case 'hop':
      newRecipe.spices.push(values);
      break;
    case 'yeast':
      newRecipe.yeast.push(values);
      break;
  }

  return newRecipe;
};

export const computeGrainWeight = (fermentables: Fermentable[]) => {
  const grainFermentables = _.filter(fermentables, fermentable => lowerCaseIncludes(fermentable.type, 'grain'));

  return _.sumBy(grainFermentables, 'weight') || 0;
};

const computeCarbVolume = (recipe: Recipe) => {
  if (/stout|porter/i.test(recipe.name)) {
    return 1.85;
  }

  if (/lambic|wheat/i.test(recipe.name)) {
    return 3.3;
  }

  return 2.5;
};

const computeFermentableMCU = (fermentable: Fermentable, batchSize: number) =>
  (fermentable.color *
    convert(fermentable.weight)
      .from('kg')
      .to('lb')) /
  convert(batchSize)
    .from('l')
    .to('gal');

const computeRecipeMCU = (fermentables: Fermentable[], batchSize: number) =>
  _.sum(_.map(fermentables, fermentable => computeFermentableMCU(fermentable, batchSize)));

const computeRecipeColor = (fermentables: Fermentable[], batchSize: number) => {
  const recipeMCU = computeRecipeMCU(fermentables, batchSize);

  return 1.4922 * Math.pow(recipeMCU, 0.6859);
};

const computeFermentableEfficiency = (fermentable: Fermentable, steepEfficiency: number, mashEfficiency: number) =>
  computeFermentableUse(fermentable) === 'steep'
    ? steepEfficiency / 100.0
    : computeFermentableUse(fermentable) === 'mash'
    ? mashEfficiency / 100.0
    : 1.0;

const computeFermentableOG = (
  fermentable: Fermentable,
  steepEfficiency: number,
  mashEfficiency: number,
  batchSize: number,
) => {
  const efficiency = computeFermentableEfficiency(fermentable, steepEfficiency, mashEfficiency);
  const gu = computeFermentableGU(fermentable, batchSize) * efficiency;

  return gu / 1000.0;
};

const computeRecipeOG = (
  fermentables: Fermentable[],
  steepEfficiency: number,
  mashEfficiency: number,
  batchSize: number,
) =>
  1 +
  _.sum(
    _.map(fermentables, fermentable => computeFermentableOG(fermentable, steepEfficiency, mashEfficiency, batchSize)),
  );

const computeTimelineMapFermentables = (
  fermentables: Fermentable[],
  type: RecipeType,
  steepEfficiency: number,
  mashEfficiency: number,
  batchSize: number,
) => {
  const initialTimelineMapFermentables: TimelineMap['fermentables'] = { mash: [], steep: [], boil: [], boilEnd: [] };

  const fermentablesWithGravity = _.map(fermentables, fermentable => ({
    fermentable,
    gravity: computeFermentableOG(fermentable, steepEfficiency, mashEfficiency, batchSize) * 1000,
  }));

  const newTimelineMapFermentables = _.groupBy(fermentablesWithGravity, ({ fermentable }) =>
    computeFermentableUse(fermentable, type),
  );

  return mergeObjects(initialTimelineMapFermentables, newTimelineMapFermentables);
};

const groupSpicesByUse = (spices: Spice[]) =>
  _.groupBy(spices, spice => {
    if (spice.use.toLowerCase() === 'boil') {
      return 'boilSpices';
    }
    if (spice.use.toLowerCase() === 'mash') {
      return 'mashSpices';
    }
    if (computeIsDrySpice(spice)) {
      return 'drySpices';
    }

    return 'otherSpices';
  });

export const calculateRecipe = (oldRecipe: Recipe) => {
  const recipe = _.cloneDeep(oldRecipe);

  // A map of various ingredient values used to generate the timeline steps below.
  recipe.timelineMap = {
    fermentables: computeTimelineMapFermentables(
      recipe.fermentables,
      recipe.type,
      recipe.steepEfficiency,
      recipe.mashEfficiency,
      recipe.batchSize,
    ),
    times: {},
    drySpice: {},
    yeast: recipe.yeast,
  };

  // Calculate properties dependant on fermentables
  recipe.color = computeRecipeColor(recipe.fermentables, recipe.batchSize);
  recipe.og = computeRecipeOG(recipe.fermentables, recipe.steepEfficiency, recipe.mashEfficiency, recipe.batchSize);

  const earlyFermentables = _.filter(recipe.fermentables, fermentable => !fermentable.late);
  const earlyOg = computeRecipeOG(earlyFermentables, recipe.steepEfficiency, recipe.mashEfficiency, recipe.boilSize);

  const fermentablesPrice = _.sum(_.map(recipe.fermentables, computeFermentablePrice));

  // calculate properties dependent on yeast
  const yeastsPrice = _.sum(_.map(recipe.yeast, computeYeastPrice));

  const attenuation = _.maxBy(recipe.yeast, 'attenuation').attenuation;
  recipe.fg = recipe.og - ((recipe.og - 1.0) * attenuation) / 100.0;

  recipe.abv = ((1.05 * (recipe.og - recipe.fg)) / recipe.fg / 0.79) * 100.0;

  // Gravity degrees plato approximations
  recipe.ogPlato = convertSpecificGravityToPlato(recipe.og);
  recipe.fgPlato = convertSpecificGravityToPlato(recipe.fg);

  recipe.realExtract = 0.1808 * recipe.ogPlato + 0.8192 * recipe.fgPlato;
  recipe.abw = (0.79 * recipe.abv) / recipe.fg;
  recipe.calories = Math.max(
    0,
    (6.9 * recipe.abw + 4.0 * (recipe.realExtract - 0.1)) * recipe.fg * recipe.servingSize * 10,
  );

  const carbVolume = recipe.bottlingPressure || computeCarbVolume(recipe);

  recipe.kegTemp = recipe.bottlingTemp || GLOBALS.REFRIGERATOR_TEMP;
  const kegTempF = convert(recipe.kegTemp)
    .from('C')
    .to('F');

  // Using brewcalc's kegPressure calculation: https://github.com/brewcomputer/brewcalc/blob/master/lib/brewcalc.js
  recipe.kegPressure = Math.max(
    0,
    -16.6999 -
      0.0101059 * kegTempF +
      0.00116512 * kegTempF * kegTempF +
      0.173354 * kegTempF * carbVolume +
      4.24267 * carbVolume -
      0.0684226 * carbVolume * carbVolume,
  );

  const bottlingTempF = convert(recipe.bottlingTemp || GLOBALS.ROOM_TEMP)
    .from('C')
    .to('F');
  recipe.primingCornSugar =
    0.015195 * 5 * (carbVolume - 3.0378 + 0.050062 * bottlingTempF - 0.00026555 * bottlingTempF * bottlingTempF);
  recipe.primingSugar = recipe.primingCornSugar * 0.90995;
  recipe.primingHoney = recipe.primingCornSugar * 1.22496;
  recipe.primingDme = recipe.primingCornSugar * 1.33249;

  // Calculate spices
  const spicesPrice = _.sum(_.map(recipe.spices, computeSpicePrice));
  const { boilSpices, mashSpices, drySpices } = groupSpicesByUse(recipe.spices);

  const mashSpiceTimelineFermentables = _.map(
    mashSpices,
    (spice): TimelineFermentable => ({
      fermentable: {
        type: 'Grain',
        name: spice.name,
        weight: spice.weight,
        color: null,
        yield: null,
      },
      gravity: 0,
    }),
  );

  recipe.timelineMap.fermentables.mash = _.concat(
    _.toArray(recipe.timelineMap.fermentables.mash),
    mashSpiceTimelineFermentables,
  );

  const timelineBoilSpices = _.map(boilSpices, spice => ({
    spice,
    bitterness: computeSpiceBitterness(spice, recipe.ibuMethod, earlyOg, recipe.batchSize),
  }));
  const spicesBitterness = _.sumBy(timelineBoilSpices, 'bitterness');
  recipe.ibu = spicesBitterness;
  recipe.timelineMap.times = _.groupBy(timelineBoilSpices, ({ spice }) => spice.time);

  const timelineDrySpices = _.map(drySpices, spice => ({
    spice,
    bitterness: 0,
  }));
  recipe.timelineMap.drySpice = _.groupBy(timelineDrySpices, ({ spice }) => spice.time);

  recipe.price = spicesPrice + fermentablesPrice + yeastsPrice;

  // Calculate bitterness to gravity ratios
  recipe.buToGu = recipe.ibu / (recipe.og - 1.0) / 1000.0;

  // http://klugscheisserbrauerei.wordpress.com/beer-balance/
  const rte = (0.82 * (recipe.fg - 1.0) + 0.18 * (recipe.og - 1.0)) * 1000.0;
  recipe.bv = (0.8 * recipe.ibu) / rte;

  return recipe;
};
