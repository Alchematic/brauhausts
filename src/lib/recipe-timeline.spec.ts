import testRecipe2 from '../test-recipe-2.json';
import testRecipe3 from '../test-recipe-3.json';
import { calculateRecipe, createRecipe, Recipe } from './recipe';
import { computeRecipeTimeline } from './recipe-timeline';

describe('computeRecipeTimeline', () => {
  it('should return the same timeline that brauhaus returns for realRecipe', () => {
    const recipe = createRecipe(testRecipe2.original as Recipe);
    const calculatedRecipe = calculateRecipe(recipe);

    const result = computeRecipeTimeline(calculatedRecipe);

    expect(result).toEqual(testRecipe2.timeline);
  });

  it('should return the same timeline that brauhaus returns for realRecipe3', () => {
    const recipe = createRecipe(testRecipe3.original as Recipe);
    const calculatedRecipe = calculateRecipe(recipe);

    const result = computeRecipeTimeline(calculatedRecipe);

    expect(result).toEqual(testRecipe3.timeline);
  });
});
