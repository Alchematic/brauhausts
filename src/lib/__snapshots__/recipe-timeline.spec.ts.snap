// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeRecipeTimeline should return the same timeline that brauhaus returns for realRecipe2 1`] = `
Array [
  Object {
    "duration": 4,
    "instructions": "Heat 2.0l to 68째C (about 4 minutes).",
    "phase": "steep",
    "time": 0,
  },
  Object {
    "duration": 20,
    "instructions": "Add 0.23kg of Pilsner malt (steeped) (1.7 GU), 0.23kg of Caramunich (steeped) (1.7 GU), 0.11kg of Aromatic (steeped) (0.8 GU) and steep for 20 minutes.",
    "phase": "steep",
    "time": 4,
  },
  Object {
    "duration": 28,
    "instructions": "Top up the wort to 11.4l and heat to a rolling boil (about 28 minutes).",
    "phase": "top-up",
    "time": 24,
  },
  Object {
    "duration": 60,
    "instructions": "Add 2.27kg of Munich liquid extract (35.0 GU), 1.36kg of Wheat liquid extract (21.0 GU), 21.2615166549g of Aramis hops, 7.08717221828g of Aramis hops",
    "phase": "boil",
    "time": 52,
  },
  Object {
    "duration": 20,
    "instructions": "Flame out. Begin chilling to 20째C and aerate the cooled wort (about 20 minutes).",
    "phase": "chill",
    "time": 112,
  },
  Object {
    "duration": 20160,
    "instructions": "Pitch WLP560 - Classic Saison Ale Blend and seal the fermenter. You should see bubbles in the airlock within 24 hours.",
    "phase": "yeast",
    "time": 132,
  },
  Object {
    "duration": 0,
    "instructions": "Prime and bottle about 53 bottles.",
    "phase": "bottle",
    "time": 20292,
  },
  Object {
    "duration": 20160,
    "instructions": "Age at 20C for 14 days.",
    "phase": "aging",
    "time": 20292,
  },
  Object {
    "duration": 0,
    "instructions": "Relax, don't worry and have a homebrew!",
    "phase": "drink",
    "time": 40452,
  },
]
`;

exports[`computeRecipeTimeline should return the same timeline that brauhaus returns for realRecipe3 1`] = `
Array [
  Object {
    "duration": 4,
    "instructions": "Heat 2.0l to 68째C (about 4 minutes).",
    "phase": "steep",
    "time": 0,
  },
  Object {
    "duration": 20,
    "instructions": "Add 0.50kg of Caramel 40L (3.6 GU) and steep for 20 minutes.",
    "phase": "steep",
    "time": 4,
  },
  Object {
    "duration": 24,
    "instructions": "Top up the wort to 10.0l and heat to a rolling boil (about 24 minutes).",
    "phase": "top-up",
    "time": 24,
  },
  Object {
    "duration": 60,
    "instructions": "Add 4.00kg of Extra pale extract (57.9 GU), 28.349556839727484g of Cascade hops",
    "phase": "boil",
    "time": 48,
  },
  Object {
    "duration": 20,
    "instructions": "Flame out. Begin chilling to 20째C and aerate the cooled wort (about 20 minutes).",
    "phase": "chill",
    "time": 108,
  },
  Object {
    "duration": 14400,
    "instructions": "Pitch Wyeast 3052 and seal the fermenter. You should see bubbles in the airlock within 24 hours.",
    "phase": "yeast",
    "time": 128,
  },
  Object {
    "duration": 7200,
    "instructions": "Move to secondary fermenter for 5 days.",
    "phase": "ferment",
    "time": 14528,
  },
  Object {
    "duration": 0,
    "instructions": "Prime and bottle about 56 bottles.",
    "phase": "bottle",
    "time": 21728,
  },
  Object {
    "duration": 20160,
    "instructions": "Age at 20C for 14 days.",
    "phase": "aging",
    "time": 21728,
  },
  Object {
    "duration": 0,
    "instructions": "Relax, don't worry and have a homebrew!",
    "phase": "drink",
    "time": 41888,
  },
]
`;
